bridge-fungible-token/src/utils.sw

```diff
@@ -173,56 +167,47 @@ pub fn parse_message_data(msg_idx: u8) -> MessageData {
+    msg_data.l1_asset = input_message_data(msg_idx, 8).into();
+    msg_data.from = input_message_data(msg_idx, 8 + 8).into();
+    msg_data.to = Address::from(input_message_data(msg_idx, 8 + 8 + 8).into());
+    msg_data.amount = input_message_data(msg_idx, 8 + 8 + 8 + 8).into();
     msg_data
 }
 
-/// Encode the data to be passed out of the contract when sending a message
-pub fn encode_data(to: b256, amount: b256) -> Vec<u64> {
-    let mut data = Vec::with_capacity(13);
-    let (recip_1, recip_2, recip_3, recip_4) = decompose(to);
-    let (token_1, token_2, token_3, token_4) = decompose(LAYER_1_TOKEN);
-    let (amount_1, amount_2, amount_3, amount_4) = decompose(amount);
-
-    // start with the function selector
-    data.push(FINALIZE_WITHDRAWAL_SELECTOR + (recip_1 >> 32));
-
-    // add the address to recieve coins
-    data.push((recip_1 << 32) + (recip_2 >> 32));
-    data.push((recip_2 << 32) + (recip_3 >> 32));
-    data.push((recip_3 << 32) + (recip_4 >> 32));
-    data.push((recip_4 << 32) + (token_1 >> 32));
-
-    // add the address of the L1 token contract
-    data.push((token_1 << 32) + (token_2 >> 32));
-    data.push((token_2 << 32) + (token_3 >> 32));
-    data.push((token_3 << 32) + (token_4 >> 32));
-    data.push((token_4 << 32) + (amount_1 >> 32));
-
-    // add the amount of tokens
-    data.push((amount_1 << 32) + (amount_2 >> 32));
-    data.push((amount_2 << 32) + (amount_3 >> 32));
-    data.push((amount_3 << 32) + (amount_4 >> 32));
-    data.push(amount_4 << 32);
-    data
-}
+fn copy_bytes(dest: raw_ptr, src: raw_ptr, len: u64, offset: u64) {
 
-/// Get the data of a message input
-// TODO: [std-lib] replace with 'input_message_data'
-pub fn input_message_data<T>(index: u64, offset: u64) -> T {
-    let data = __gtf::<raw_ptr>(index, GTF_INPUT_MESSAGE_DATA);
-    let data_with_offset = data.add(offset / 8);
-    data_with_offset.read::<T>()
+    asm(dst: dest, src: src, len: len) {
+            mcp dst src len;
+        };
 }
 
-/// Get the sender of the input message at `index`.
-// TODO: [std-lib] replace with 'input_message_sender'
-pub fn input_message_sender(index: u64) -> Address {
-    Address::from(__gtf::<b256>(index, GTF_INPUT_MESSAGE_SENDER))
+/// Encode the data to be passed out of the contract when sending a message
+pub fn encode_data(to: b256, amount: b256) -> Bytes {
+    // cap is 4 + 20 + 20 + 32 = 76
+    let mut data = Bytes::with_capacity(76);
+    let padded_to_bytes = Bytes::from(to);
+    let padded_token_bytes = Bytes::from(LAYER_1_TOKEN);
+    let amount_bytes = Bytes::from(amount);
+
+    // first, we push the selector 1 byte at a time
+    // the function selector for finalizeWithdrawal on the L1ERC20Gateway contract:
+    // finalizeWithdrawal(address,address,uint256) = 0x53ef1461
+    data.push(0x53u8);
+    data.push(0xefu8);
+    data.push(0x14u8);
+    data.push(0x61u8);
+
+    // next, we copy the least significant 20 bytes of the `to` address:
```

Comment:
The EVMs ABI encoding expects addresses to be zero padded to a full 32 bytes. We need to copy the whole 32 bytes for both "to" and "LAYER_1_TOKEN".

Reply:
Ah, I see. That's an easy enough change to make.